rfcs/rdm-0064-custom-fields.md

```diff
@@ -0,0 +1,440 @@
+
+#### Adding or overriding UI wigets
+
+In the previous section, the `ui_widget` field was mentioned. Several of them
+are already available with InvenioRDM, however, in some cases it might be
+desired to change their behaviour (override) or use a totally new widget.
+
+The webpack loading order for the custom fields is the following:
+
+1. Try to import the widget from the RDM instance. The importer checks the cookiecutter `/assets/templates` folder. That folder is copied/symlinked when the assets are built.
+2. If the widget was not found, then it tries to import it from the local module's path.
+3. If the component is still not found, then it raises an error.
+
+With the above mechanism, instance administrators can add new files/UI widgets that can then be referenced in the config. For example, if a instance administrator wants to add a new UI widget called `MyCustomWidget` then they need to:
+
+1. Go to `/my-site/assets/templates` folder
+2. Create a new file called `MyCustomWidget.js`
+3. Use a `default` export, to export the widget
+
+```javascript
+/my-site/assets/templates/custom_fields/MyCustomWidget.js
+
+import React, { Component } from "react";
+import PropTypes from "prop-types";
+
+import { FieldLabel, TextAreaField } from "react-invenio-forms";
+
+export default class MyCustomField extends Component {
+  render() {
+    const { fieldPath, label, icon, placeholder, description } = this.props;
+
+    return (
+      <>
+        <TextAreaField
+          fieldPath={fieldPath}
+          label={<FieldLabel htmlFor={fieldPath} icon={icon} label={label} />}
+          placeholder={placeholder}
+        />
+        {description && <label className="helptext">{description}</label>}
+      </>
+    );
+  }
+}
+```
+
+4. Then the instance administrator can reference that widget in the config:
+
+```python
+RDM_RECORDS_CUSTOM_FIELDS_UI = [{
+    "section": _("CERN Fields"),
+    "fields": [{
+        ...
+        "ui_widget": "MyCustomField",
+        ...
+    }]
+}]
+```
+
+In a similar manner, instance administrators can redefine in their RDM instance an existing UI widget and override completely its behaviour. For this they would need to create a file with the same name as the component they want to override.
+
+### Overriding how fields will be displayed in the record's / community's detail page
+
+The custom fields will be displayed by default in a _to-be-defined_ place in the record's landing page and community's detail page. The default place will be decided to fit the up-to-date UX design of these pages. That said, every institution has their own UI design and it is already possible to override the default display by overriding the Jinja templates of these pages.
+
+### Adding a custom field as facet in the search page
+
+Search filtering or facetting would be configured as it was before, using the and `RDM_FACETS` variable.
+
+```python
+RDM_FACETS = {  # To add search filtering
+    "experiment": CFTermsFacet()
```

Comment:
We are using `<Type>CF` above, but `CF<Type>Facet` here... is that ok?

Reply:
Yes, we do not "automagically" add facets. We use the existing RDM mechanism, which means using `RDM_FACETS` config var and `...Facet` classes. I do agree this could be abstracted, we can open an issue for a second iteration.